rasa/core/trackers.py

```diff
@@ -420,11 +432,67 @@ def undo_till_previous(event_type, done_events):
                 # the `action_listen` action right before it).
                 undo_till_previous(UserUttered, applied_events)
                 undo_till_previous(ActionExecuted, applied_events)
+            elif (
+                isinstance(event, ActionExecuted)
+                and event.action_name in form_names
+                and not self._upcoming_rejection(idx)
+                and not self._first_loop_execution_or_previous_rejection(
+                    event.action_name, applied_events
+                )
+            ):
+                self.undo_till_previous_loop_execution(
+                    event.action_name, applied_events
+                )
+            elif isinstance(event, ActionExecutionRejected):
+                # Undo previously executed action
+                undo_till_previous(ActionExecuted, applied_events)
+                applied_events.append(event)
```

Comment:
why we didn't do it before? how it was handled?

Reply:
I think the `ActionExecuted` event is not added in this case, so you revert the wrong action